docs/configuration/module-config/serial.mdx

```diff
@@ -294,3 +294,170 @@ previousState = currentState; // update the previous state
+
+      if (buttonState == LOW) { // if the button pin is still LOW after debounce delay, that means the detected change was a real press, so we send the relay toggle command. If not, then we just assume the state change was noise and do nothing.
+
+        sendRelayToggleCommand();
+
+      }
+
+    }
+
+  }
+
+}
+
+void sendRelayToggleCommand() {
+
+  Serial.println(relayToggleCode);
+
+}
+
+```
+
+#### Arduino Mini Pro / relay Wiring
+
+![image](/img/modules/Serial/relay.png)
+
+- T-BEAM RX PIN 13 to TX PIN on the ARDUINO MINI
+- T-BEAM TX PIN 14 to RX PINon the ARDUINO MINI
+- T-BEAM PIN 3.3V to 3.3V PIN on the ARDUINO MINI
+- T-BEAM PIN GND to GND PIN on the ARDUINO MINI
+- ARDUINO MINI PIN 2 to SIGNAL PIN on the RELAY
+- ARDUINO MINI PIN GND to the GND on the RELAY
+- ARDUINO MINI PIN VCC to the VCC on the RELAY
+
+#### Arduino Mini Pro /relay Code
+
+```cpp
+//define sketch constants (gpios, baud rate and other constants)
+#define RELAY_PIN 2
+#define BAUD_RATE 9600
+
+int currentRelayState;
+const char* relayToggleCode = "relay1toggle";
+const char* relayOnCode = "on";
+const char* relayOffCode = "off";
+const unsigned int MAX_INPUT = 30;
+
+void setup() {
+  Serial.begin(BAUD_RATE);
+  pinMode(RELAY_PIN, OUTPUT);
+  digitalWrite(RELAY_PIN, LOW);
+  currentRelayState = LOW;
+}
+
+void loop() {
+  while (Serial.available() > 0) {
+    processIncomingByte(Serial.read());
+  }
+}
+
+void processIncomingByte(const byte inByte) {
+  static char input_line[MAX_INPUT];
+  static unsigned int input_pos = 0;
+
+  switch (inByte) {
+    case '\n':
+      input_line[input_pos] = 0;
+      checkStringForPasscode(input_line);
+      input_pos = 0;
+      break;
+    case '\r':
+      break;
+    default:
+      if (input_pos < (MAX_INPUT - 1))
+        input_line[input_pos++] = inByte;
+      break;
+  }
+}
+
+void checkStringForPasscode(const char* data) {
+  String inputString = data;
+
+  if (inputString.indexOf(relayToggleCode) != -1) {
+    if (currentRelayState == LOW) {
+      digitalWrite(RELAY_PIN, HIGH);
+      currentRelayState = HIGH;
+      const char* confirmationMessage = "Relay 1 has received the command and is now ON";
+      Serial.println(confirmationMessage);
+    } else {
+      digitalWrite(RELAY_PIN, LOW);
+      currentRelayState = LOW;
+      const char* confirmationMessage = "Relay 1 has received the command and is now OFF";
+      Serial.println(confirmationMessage);
+    }
+  } else if (inputString.indexOf(relayOnCode) != -1) {
+    digitalWrite(RELAY_PIN, HIGH);
+    currentRelayState = HIGH;
+    const char* confirmationMessage = "Relay 1 has received the command and is now ON";
+    Serial.println(confirmationMessage);
+  } else if (inputString.indexOf(relayOffCode) != -1) {
+    digitalWrite(RELAY_PIN, LOW);
+    currentRelayState = LOW;
+    const char* confirmationMessage = "Relay 1 has received the command and is now OFF";
+    Serial.println(confirmationMessage);
+  }
+}
```

Comment:
```suggestion
//define sketch constants (pins, baud rate and other constants)
#define RELAY_PIN 2
#define BAUD_RATE 9600

int currentRelayState; //this variable can be checked to see the state of the relay pin and must be manually updated any time you change the state of the relay pin with digitalWrite(). HIGH means active coil, LOW means inactive coil

//these "code" variables are the commands to control the relay, they can be whatever you want them to be, but make sure they are not too simple like "on" or "off" or you will accidentally activate/deactivate your relay while sending basic text conversation over the mesh. For example, if the relayOffCode is just "off" and you send "Hey Steve, what time are you off work?", you've just turned off your relay accidentally because the string "off" was in your message.
const char* relayToggleCode = "relay1toggle";
const char* relayOnCode = "relay1on";
const char* relayOffCode = "relay1off";

// the MAX_INPUT variable defines the size of the incoming message buffer that checkStringForPasscode() examines to see if it contains any of the "codes" above, this value can be whatever you want but smaller numbers will get processed faster. As long as you send the right command within the first 30 bytes (characters) of a message from another node, your command will be seen by the Arduino and processed. (it's actually more like the first 24 bytes because you have to account for the first 6 characters that the Serial Module prepends to each incoming message in the form of the sending node's shortname (4 characters) which is always followed by ": " (another 2 characters, so 6 in total). 
const unsigned int MAX_INPUT = 30;

void setup() {
  Serial.begin(BAUD_RATE); //enable UART so Tbeam and Arduino can talk to each other
  pinMode(RELAY_PIN, OUTPUT);
  
  //the lines below were written so that the relay always starts off in the inactive state on the Arduino bootup.  But if you had some way of writing variables to non-volatile memory like flash memory you could save the most recent state of the relay to flash and set the relay to that state every time the Arduino is powered on. Just something to consider. 
  digitalWrite(RELAY_PIN, LOW); 
  currentRelayState = LOW;
}

void loop() {
  //the main loop() in this sketch is very simple, it just checks to see if any new data is in serial buffer and then if there is, it sends the next available byte (character) to the processIncomingByte() function
  while (Serial.available() > 0) {
    processIncomingByte(Serial.read());
  }
}

//Credit for the following function goes to Nick Gammon 
//This function takes the incoming bytes from the serial buffer and stores them in a string or char array. Everytime the function is called it adds the new byte or character to the input_line array until the array has the MAX_INPUT number of elements or until a new-line character '\n' is received. Carriage returns '\r' are ignored. When the '\n\' character is received it is appended to the array and the whole array is passed to the checkStringForPasscode() function.
void processIncomingByte(const byte inByte) {
  static char input_line[MAX_INPUT];
  static unsigned int input_pos = 0;

  switch (inByte) {
    case '\n':
      input_line[input_pos] = 0;
      checkStringForPasscode(input_line);
      input_pos = 0;
      break;
    case '\r':
      break;
    default:
      if (input_pos < (MAX_INPUT - 1))
        input_line[input_pos++] = inByte;
      break;
  }
}

//this function does the exciting stuff that you can see with your eyes. It stores the incoming char array as a String object (yes with a capital S, not to be confused with a string variable), so that we can process the data with the .indexOf() function. The indexOf() function is used to find a, let's call it a "sub-string", within the String object. If it finds the sub-string, the function returns the index of that sub-string. If the sub-string does not exist within the String object, the function returns a -1. So as long as .indexOf() does NOT return a -1, we know that the sub-string exists in the String.
void checkStringForPasscode(const char* data) {
  
  //stores the incoming message as a String object
  String inputString = data;

  if (inputString.indexOf(relayToggleCode) != -1) { //if relayToggleCode exists in inputString, toggle the relay...
    
    //to toggle the relay, we first check "currentRelayState". If it is LOW, then we toggle the relay HIGH, and if it is HIGH, we set it back to LOW. We also respond to the mesh with a confirmationMessage or an acknowledgement that a command was received and the Arduino has reacted to that command appropriately. This way you can know for sure that your relay is working properly even if you are miles away.
    if (currentRelayState == LOW) {
      digitalWrite(RELAY_PIN, HIGH);
      currentRelayState = HIGH;
      Serial.println("Relay 1 has received the command and is now ON");
    } else {
      digitalWrite(RELAY_PIN, LOW);
      currentRelayState = LOW;
      Serial.println("Relay 1 has received the command and is now OFF");
    }
  
  } else if (inputString.indexOf(relayOnCode) != -1) { //if the message contains "relay1on"...
  
    digitalWrite(RELAY_PIN, HIGH); //turn on the relay
    currentRelayState = HIGH; //update the currentRelayState variable
    Serial.println("Relay 1 has received the command and is now ON"); //reply to the mesh with acknowledgement
  
  } else if (inputString.indexOf(relayOffCode) != -1) { //if the message contains "relay1off"...
    
    digitalWrite(RELAY_PIN, LOW); //turn off the relay
    currentRelayState = LOW; //update the currentRelayState variable
    Serial.println("Relay 1 has received the command and is now OFF"); // reply to the mesh with acknowledgement
  
  }
  
}
```

Reply:
please read over my proposed comments @pdxlocations and let me know if anything needs changed, thank you

Reply:
@pdxlocations also, is there any way to preview what this code would look like on the actual site? i'm assuming you guys use Vercel to build pre-merge previews, is there anyway for me to view that too or is it only available fore the core dev team?

Reply:
Yes, instructions are at: https://meshtastic.org/docs/development/documentation/publish#publish-to-vercel

You'll need a GitHub account and Vercel account.  One thing that wasn't clear when I started using Vercel is that you have to create a branch of your fork of the site then submit a pull request to your own master branch (not the meshtastic/meshtastic master) for a shareable preview deployment to be generated.  Once you're happy with your work, THEN submit the PR to meshtastic/meshtastic. 

Alternatively you can host a preview locally but then you can't share a preview link:
https://meshtastic.org/docs/development/documentation/local-dev

I use both methods:  First creating a local preview then PR my own fork, then PR the main repo.  This workflow works for me but it's kinda slow.  YMMV.

I'll look at your code when I get a chance.  I'd like to test it out on some hardware if I can find the time too!

Reply:
@philmanofsky Sorry I've been busy with some family things.  I tried to get this code running at home but I'm having problems getting the serial connection going on my T-Beam.  It would probably help if I started by soldering on headers to actually attach the wires instead of just sticking them in the holes, lol.  

I did notice that comments are extremely helpful but perhaps a few might be a bit long.  They also don't format well when I paste them into the Arduino IDE.  They just show up as one long line.  Might be a problem with my settings.  I'll have to investigate further.

Are you planning to create a preview of the changes with Vercel?  That would make this pull easier examine.

Reply:
> Preview should be available now, however, I do not think your proposed changes will be shown in the preview. I believe it builds only once a commit is added. To view that you would need to follow @pdxlocations's advice. That is also what I use for my personal development.

@rcarteraz  Could hitting the "commit suggestion" button force a new vercel preview? I still have a lot of git to learn.

![Screenshot 2023-07-30 at 9 53 56 PM](https://github.com/meshtastic/meshtastic/assets/117498748/a47df7e6-4632-4388-9f02-f6e66ce3a599)